### PR TITLE
[Issue #125] Enable DatadogCore and DatadogLogs to compile on watchOS platform

### PR DESCRIPTION
**Pull request by jfiser-paylocity**
### What and why?
Related to issue https://github.com/DataDog/dd-sdk-ios/issues/125.

This PR resolves compilation errors for watchOS platform in the following targets:
- DatadogInternal
- DatadogCore
- DatadogLogs

### How?
Minimum required version of watchOS 7 has been added to SPM/Cocoapods/Carthage definitions. The biggest issue on lower version is the lack of system notifications (didBecomeActive, didEnterBackground, willResignActive, willEnterForeground).

There are couple of limitations on watchOS platform overall, however I don’t see these as blockers:
- No background task - Logs will still sync when the app is on the foreground.
- HTTP header `Proxy-Authorization` is not set - Could be mitigated by using String static values
- Kronos library always returns local time

To test this, you can temporarily add Apple Watch under Supported Destinations in respective targets. However it might be good to ensure at compilation time that watchOS platform does not break in the future.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests for Session Replay
- [ ] Run smoke tests
- [ ] Run tests for `tools/`

### Example
I tried to integrate fork changes directly in the POC project of mine to validate it works. See below is the example log opened in web Datadog dashboard.

image

### CODE REVIEW

File: DatadogCore/Sources/Core/Context/ApplicationStatePublisher.swift [hunk 1]
```diff
@@ -7,12 +7,19 @@
...
+import WatchKit
+#endif
 
 internal final class ApplicationStatePublisher: ContextValuePublisher {
     typealias Snapshot = AppStateHistory.Snapshot
 
-    private static var currentApplicationState: UIApplication.State {
+    private static var currentApplicationState: ApplicationState {
+        #if canImport(WatchKit)
+        WKExtension.shared().applicationState
```
Comment (maciejburda) [inline DatadogCore/Sources/Core/Context/ApplicationStatePublisher.swift:19]: probably worth introducing `.dd` accessor, similarly to `UIApplication` below

File: DatadogCore/Sources/Core/Upload/FeatureUpload.swift [hunk 2]
```diff
@@ -38,7 +38,11 @@ internal struct FeatureUpload {
         let backgroundTaskCoordinator: BackgroundTaskCoordinator?
         switch (backgroundTasksEnabled, isRunFromExtension) {
         case (true, false):
+            #if os(watchOS)
+            backgroundTaskCoordinator = nil
```
Comment (maciejburda) [inline DatadogCore/Sources/Core/Upload/FeatureUpload.swift:42]: There's a chance it would work with `ExtensionBackgroundTaskCoordinator`

**Review by maciejburda — COMMENTED** [review]
Interesting and valuable contribution! 🙇

It's quite complex to test this, but I'll try to schedule some time next week to give it a closer look.

For now I enabled CI run and gave it a first pass.

File: DatadogInternal/Sources/Utils/WatchKitExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
+//
+//  WatchKitExtensions.swift
+//  Datadog
+//
+//  Created by Jakub Fiser on 26.06.2024.
+//  Copyright © 2024 Datadog. All rights reserved.
+//
```
Comment (maciejburda) [inline DatadogInternal/Sources/Utils/WatchKitExtensions.swift:7]: Let's change to default project header 🙏

**Review by maciejburda — COMMENTED** [review]
In general it LGTM!

I think we need to consider writing a smoke test for watch kit build (the same way we do it for catalyst and macOS). With this we'll prevent from breaking the build in the future releases.

**Review by ncreated — CHANGES_REQUESTED** [review]
@jfiser-paylocity Thank you very much for your contribution! Your effort in identifying all the necessary areas for updates is impressive and much appreciated 🏅.

We will take some time to thoroughly test these changes to ensure basics are fine and nothing crucial is missed.

In the meantime, could you please update your branch with our latest `develop` and conduct a round of minimal testing? It seems that some tests are currently failing:
```
make spm-build-ios ✅
make spm-build-tvos ✅
make spm-build-visionos ✅
make spm-build-macos ✅

make test-ios-all ❌
make test-tvos-all ❌
```
The failures in the last two commands (unit tests) are likely due to the omission of the two new `DatadogInternal` files in `Datadog.xcodeproj`. Including these files is crucial for building Carthage artifacts and deploying XCFrameworks from the SDK. Please address address this issue. It is is the only blocker I can see for now.
```
Testing failed:
	Cannot find type 'DeviceProtocol' in scope
```

**Comment by jfiser-paylocity** [comment]
@maciejburda @ncreated Thank you for your response. All the mentioned issues should be resolved now. Please let me know if there is anything I may have missed.

**Review by ncreated — APPROVED** [review]
@jfiser-paylocity Thanks for updates and enabling spm tests for watchOS 👌 🎯! I did more validations on our side and it looks good. Like I said before this contribution looks very exhaustively and high quality 🏅 so we should be with shipping it into our next releases.

We will proceed with this PR by first merging it to our own branch, so we can run full CI pipeline and do last polishing.

**Comment by johnHackworth** [comment]
Hey @jfiser-paylocity 👋 Just wanted to close the loop — your contribution here was a key stepping stone toward the full watchOS support that shipped in **[v3.10.0](https://github.com/DataDog/dd-sdk-ios/releases/tag/3.10.0)**! We now officially support RUM, Crash Reporting, Logs, Trace, and Feature Flags on watchOS. Thank you for your work on this 🙏

**Pull request merged**